src/config.rs

```diff
@@ -37,188 +39,55 @@ impl Default for Config {
-    pub fn cache_capacity(&self, cache_capacity: usize) -> Config {
-        let mut ret = self.clone();
-        ret.cache_capacity = cache_capacity;
-        ret
-    }
+macro_rules! builder {
+    ($(($name:ident, $get:ident, $set:ident, $t:ty, $desc:expr)),*) => {
+        $(
+            #[doc="Get "]
+            #[doc=$desc]
+            pub fn $get(&self) -> $t {
+                self.$name.clone()
+            }
 
-    /// Builder, set whether the system uses the OS pagecache for
-    /// compressed pages.
-    pub fn use_os_cache(&self, use_os_cache: bool) -> Config {
-        let mut ret = self.clone();
-        ret.use_os_cache = use_os_cache;
-        ret
-    }
+            #[doc="Set "]
+            #[doc=$desc]
+            pub fn $set(&mut self, to: $t) {
+                self.$name = to;
+            }
 
-    /// Builder, set whether the system compresses data written to
-    /// secondary storage with the zstd algorithm.
-    pub fn use_compression(&self, use_compression: bool) -> Config {
-        let mut ret = self.clone();
-        ret.use_compression = use_compression;
-        ret
+            #[doc="Builder, set "]
+            #[doc=$desc]
+            pub fn $name(&self, to: $t) -> Config {
+                let mut ret = self.clone();
+                ret.$name = to;
+                ret
+            }
+        )*
     }
+}
 
-    /// Builder, set the number of milliseconds between flushes.
-    pub fn flush_every_ms(&self, flush_every_ms: Option<u64>) -> Config {
-        let mut ret = self.clone();
-        ret.flush_every_ms = flush_every_ms;
-        ret
-    }
+impl Config {
+    builder!(
+        (io_bufs, get_io_bufs, set_io_bufs, usize, "number of io buffers"),
+        (io_buf_size, get_io_buf_size, set_io_buf_size, usize, "size of each io flush buffer"),
+        (blink_fanout, get_blink_fanout, set_blink_fanout, usize, "b-link node fanout"),
+        (page_consolidation_threshold, get_page_consolidation_threshold, set_page_consolidation_threshold, usize, "page consolidation threshold"),
+        (path, get_path, set_path, Option<String>, "path for the main storage file"),
+        (cache_bits, get_cache_bits, set_cache_bits, usize, "log base 2 of the number of cache shards"),
+        (cache_capacity, get_cache_capacity, set_cache_capacity, usize, "maximum size for the system page cache"),
+        (use_os_cache, get_use_os_cache, set_use_os_cache, bool, "whether to use the OS page cache"),
+        (use_compression, get_use_compression, set_use_compression, bool, "whether to use zstd compression"),
+        (flush_every_ms, get_flush_every_ms, set_flush_every_ms, Option<usize>, "number of ms between IO buffer flushes"),
+        (snapshot_after_ops, get_snapshot_after_ops, set_snapshot_after_ops, usize, "number of operations between page table snapshots"),
+        (snapshot_path_prefix, get_snapshot_path_prefix, set_snapshot_path_prefix, Option<String>, "snapshot file prefix")
+    );
```

Comment:
it's sort of weird that you have to say 3 names, instead of the macro being able to "concatenate identifiers" like just putting "set_" and "get_" in front of the first name. this is waiting on a feature called "concat_idents!" which I don't think has gotten added yet.